### PR TITLE
Use shapeHelperInferShapes template to reduce boilerplate code.

### PR DESCRIPTION
**Pull request by etiotto**
This PR leverages the template function `shapeHelperInferShapes` to reduce code "duplication" in `ShapeInference.cpp` for several ONNX operators. As an example, the following implementation of the Transpose operator `inferShapes` member function:

```
ONNXTransposeOp::inferShapes(
    std::function<void(mlir::Region &)> doShapeInference) {
  // Cannot infer shape if no shape exists.
  if (!data().getType().isa<RankedTensorType>())
    return success();

  auto elementType = data().getType().cast<ShapedType>().getElementType();
  ONNXTransposeOpAdaptor operandAdaptor(*this);
  ONNXTransposeOpShapeHelper shapeHelper(this);
  if (failed(shapeHelper.computeShape(operandAdaptor)))
    return emitError("Failed to scan Transpose parameters successfully");
  SmallVector<int64_t, 4> outputDims;
  IndexExpr::getShape(shapeHelper.dimsForOutput(), outputDims);
  getResult().setType(RankedTensorType::get(outputDims, elementType));
  
  return success();
}
```

Becomes:
```
ONNXTransposeOp::inferShapes(
    std::function<void(mlir::Region &)> doShapeInference) {
  return shapeHelperInferShapes<ONNXTransposeOpShapeHelper, ONNXTransposeOp,
      ONNXTransposeOpAdaptor>(this, elementType);
}
```

Signed-off-by: Ettore Tiotto <etiotto@ca.ibm.com>

### CODE REVIEW

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 1]
```diff
@@ -660,22 +647,13 @@ LogicalResult ONNXArgMinOp::verify() {
...
-  ONNXArgMinOpShapeHelper shapeHelper(this);
-  ONNXArgMinOpAdaptor operandAdaptor(*this);
-  if (failed(shapeHelper.computeShape(operandAdaptor)))
-    return emitError("Failed to scan ArgMin parameters successfully");
-
-  SmallVector<int64_t, 4> outputDims;
-  IndexExpr::getShape(shapeHelper.dimsForOutput(), outputDims);
-
   // ONNX spec specifies the reduced type as an int64
   Type elementType = IntegerType::get(getContext(), 64);
```
Comment (whitneywhtsang) [inline src/Dialect/ONNX/ONNXOps.cpp:654]: For consistency, 
```suggestion
  auto elementType = IntegerType::get(getContext(), 64);
```

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 2]
```diff
@@ -1139,9 +1117,11 @@ LogicalResult ONNXSequenceLengthOp::inferShapes(
 /// the shape inference interface.
 LogicalResult ONNXPReluOp::inferShapes(
     std::function<void(mlir::Region &)> doShapeInference) {
-  if (!X().getType().isa<RankedTensorType>() ||
-      !slope().getType().isa<RankedTensorType>())
-    return success();
+  ONNXPReluOpAdaptor operandAdaptor(*this);
+  for (const auto &operand : operandAdaptor.getOperands())
+    if (!hasShapeAndRank(operand))
```
Comment (whitneywhtsang) [inline src/Dialect/ONNX/ONNXOps.cpp:1122]: Use `any_of`?

File: src/Dialect/ONNX/ShapeInference/TopK.cpp [hunk 3]
```diff
@@ -48,9 +48,9 @@ LogicalResult ONNXTopKOpShapeHelper::computeShape(
       outputDims.emplace_back(XBounds.getDim(i));
   }
 
-  // There are two outputs: one for values and one for indices.
```
Comment (whitneywhtsang) [inline src/Dialect/ONNX/ShapeInference/TopK.cpp:51]: This change makes the patch no longer NFC?
Can we do this change in a different PR?

Comment (etiotto) [reply src/Dialect/ONNX/ShapeInference/TopK.cpp:51]: This is required to apply the template for the TopK operator so I prefer to keep it here. I can remove the "NFC".

File: src/Dialect/ONNX/ONNXOps.cpp (hunk 2 above)
Comment (etiotto) [reply src/Dialect/ONNX/ONNXOps.cpp]: done

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, code looks much better
